### ISSUE TITLE
Allow godot-cpp to be installable

### ISSUE THREAD

**Issue by ytnuf** [godotengine/godot-proposals#9364]
### Describe the project you are working on

Currently making a gdextension library for dammaku.  

### Describe the problem or limitation you are having in your project

Currently the only way to use godot-cpp is to use as a subproject of your project.  

I personally would prefer to use a package manager such as vcpkg.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I have made a PR that would make godot-cpp installable.

https://github.com/godotengine/godot-cpp/pull/1418

There are other PRs that also try to address this:

https://github.com/godotengine/godot-cpp/pull/1309
https://github.com/godotengine/godot-cpp/pull/1041  

This would also have the advantage of being used in a docker image as well.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

```
find_package("godot-cpp" 4.2.0 CONFIG REQUIRED)
target_link_library("my-gdextension-project" PRIVATE "godot::cpp")
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This could worked around by patching it.  However it's error-prone and would need to repeated for every package manager (conan, hunter, vcpkg, etc).  

Making the library installable, would mean it's a lot easier to port this to package managers.

### Is there a reason why this should be core and not an add-on in the asset library?

This feature is for godot-cpp.

**Comment by dsnopek** [comment]
Installing godot-cpp system-wide and dynamically linking it will create some issues with the binary compatibility model of godot-cpp.

Currently, you build your extension to statically include the version of godot-cpp and `extension_api.json` of the Godot version you are targeting, which will allow it to work with that version of Godot or later. Other extensions you use may be compiled against a different godot-cpp/`extension_api.json`, and you can use those together in the same project without problems.

If your extension were dynamically linked to a system-wide godot-cpp, and you updated the system-wide godot-cpp for a newer version of Godot, it could break existing extensions that were compiled against the older godot-cpp. So, you end up in a situation where you need to update and re-compile all your GDExtensions to work with the one godot-cpp.

**Comment by ytnuf** [comment]
Yep, this doesn't change how godot-cpp is built, so it will still be a static library.

**Comment by dsnopek** [comment]
@enetheru Does adding this feature make sense to you?

**Comment by enetheru** [comment]
@dsnopek I think if it helps some of the users then I would be asking for reasons why not, rather than why. Then seeing if the negatives can be mitigated or catered to. For instance: feature parity with scons is mandatory, so these PR's only cater to cmake which is half the work needed.

I would think that in the beginning of a developers journey, reducing the barrier to entry, or steps and complexity required to get moving is a good thing. Pre-built libraries that are easy to consume satisfy that philosophy. When they eventually upgrade to custom builds of godot which require building godot-cpp from scratch they can cross the bridge of building it themselves.

Off the top of my head I think about cross compile toolchains, version naming, and the kinds of support requests that will come from users of pre-built packages, but I know all of these things have been solved many times over in other projects so they are just technical questions to be answered, and I believe a lot of that falls to package maintainers rather than godot, so long as the mechanisms for naming and install are there I think we'll be good.

My plan is to go over these, and other PR's after the modernise PR is done, approved, and merged. I'll work with willing developers to get their PR's merged, or at least brought upto date.

**Comment by ytnuf** [comment]
As for SCons feature parity, that can be done.

**Comment by dsnopek** [comment]
> I think if it helps some of the users then I would be asking for reasons why not, rather than why.

The usual "why not": it's more code that we'll have to maintain going forward.

> As for SCons feature parity, that can be done.

Thanks for adding the SCons implementation to your PR! That makes it much more clear to me what you're trying to accomplish, since I don't know CMake very well.

It _feels_ like a lot of code for a use-case that doesn't seem super common. But I'll spend some time thinking about it, and we should probably discuss this at a GDExtension meeting too.

**Comment by dsnopek** [comment]
Discussed at the GDExtension meeting, and we don't think it makes sense to add support for this into godot-cpp's build system.

The use case described here ("including a pre-built godot-cpp in Docker images for CI") should already be possible without having to alter godot-cpp's build system, either by just building the godot-cpp directory in your project and including that in your Docker image, or by using custom scripts.

Overall, we don't want to give the impression that using godot-cpp as an installed library is a supported or recommended way to use it. godot-cpp is designed to be embedded within your GDExtension project, and this affects how many things are handled, for example, backwards compatibility and toolchain compatibility, as well as using a custom `extension_api.json` or build profile.

**Comment by loicmouton** [comment]
Has there been any discussions about it since then ? I was looking at how to install godot-cpp with CMake to be able to use it on several GDExtensions I'm working on.
The idea is to install it once, and share it to other developers that are working on the same GDExtensions and targeting the same Godot version.
As I understand it, it is the canonical way to handle this with CMake so I'm a bit surprised it was rejected.

**Comment by enetheru** [comment]
> As I understand it, it is the canonical way to handle this with CMake so I'm a bit surprised it was rejected.

I've been thinking about it on and off since it was rejected, wondering whether there might be an implementation of this proposal that respects the desires of the maintainers simultaneously with those of the users.

I've seen enough examples of cmake "installation" just meaning to generate and copy the consumable files to a specific location, typically as a configure step of another build tree to know that the workflow is very common.

**Comment by kelteseth** [comment]
@loicmouton I'm working on porting godot-cpp vcpkg to `4.5` https://github.com/godotengine/godot-cpp/issues/1882. As outlined in https://github.com/godotengine/godot-cpp/issues/1882#issuecomment-3541828971 we build vcpkg once and then share the build across all dev machines, because godot-cpp is not our only dependency. Compiling all dependencies we use takes about an hour.
